native/membrane_videocompositor/src/new_compositor.rs

```diff
@@ -0,0 +1,62 @@
+use std::sync::{Arc, Mutex};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct VideoId(usize);
+
+pub struct Frame {}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Timestamp(u64);
+
+pub struct MockScene {}
+
+#[derive(Debug, thiserror::Error)]
+pub enum CreatingCompositorError {}
+
+#[derive(Debug, thiserror::Error)]
+pub enum UpdateSceneError {}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ComposeError {}
+
+pub struct Compositor {
+    scene: MockScene,
+
+    // I keep this in an Arc<Mutex<>> so that the frame queueing mechanism can have
+    // references to this stuff to upload frames to the gpu memory. Is this a good idea?
+    // should the queue be able to do that, or would that be a bad separation of concerns?
+    //
+    // If the queue uploads the frames, this part of the compositor does not need to be
+    // concerned with stream format changes etc.
+    _wgpu: Arc<Mutex<(wgpu::Device, wgpu::Adapter)>>,
+}
+
+impl Compositor {
+    pub fn new(
+        _wgpu: Arc<Mutex<(wgpu::Device, wgpu::Queue)>>,
+    ) -> Result<Self, CreatingCompositorError> {
+        todo!()
+    }
+
+    pub fn update_scene(&mut self, scene: MockScene) -> Result<(), UpdateSceneError> {
+        self.scene = scene;
+        todo!()
+    }
+
+    /// ## Parameters:
+    ///  - `frames`: a `Vec` with a frame for each video present in the scene
+    ///  - `pts`: the pts of the composed frame. This value can be used in transformations
+    ///    that require temporal information (e.g. fade-outs)
+    pub fn compose(
+        &self,
+        _frames: Vec<(VideoId, Frame)>,
+        _pts: Timestamp,
+    ) -> Result<Frame, ComposeError> {
+        todo!()
+    }
+}
+
+// Open questions:
+//  - do we need pts for each frame? I think we don't and components that use time information
+//    should use the timestamp of the whole frame as their time reference, but I would love to
+//    hear some opinions on that
```

Comment:
I currently can't come up with a component in which having individual frames pts would be beneficial. Moreover, I think handling timestamp normalization in the queue would simplify creating new components and it's more natural for me to handle that there.

Reply:
I reckon, that in one of the earliest brainstorming of VC we decided components should not rely on the pts. Instead, if they ever need one, VC can always provide them percentages or something similar.

Reply:
So I second @WojciechBarczynski opinion

Reply:
> I reckon, that in one of the earliest brainstorming of VC we decided components should not rely on the pts. Instead, if they ever need one, VC can always provide them percentages or something similar.

This seems complicated, I'm not sure how we would pass this time information, since I think calculating it would be the responsibility of the queue. Any ideas?